### PR TITLE
Supports sharing cache_promote LRUs across multiple remap rules

### PR DESCRIPTION
**Pull request by zwoop**
This also refactors the code, and cleans things up. When merging, this should all be squashed, I left them separate for ease of review.

### CODE REVIEW

File: plugins/cache_promote/chance_policy.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      {"promoted", &promoted_id},
+      {"total_requests", &total_requests_id},
+    };
+
+    if (nullptr == remap_id) {
+      TSError("[%s] no remap identifier specified for for stats, no stats will be used", PLUGIN_NAME);
+      return false;
+    }
+
+    for (int ii = 0; ii < 3; ii++) {
```
Comment (SolidWallOfCode) [inline plugins/cache_promote/chance_policy.h:64]:
```
for ( auto && [ name, id ]: stats ) {
  if ((*id == create_stat(name, remap_identifier)) == TS_ERROR) {
    return false;
  }
}
return true;
```

File: plugins/cache_promote/lru_policy.cc [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+#include "lru_policy.h"
+
+#define MINIMUM_BUCKET_SIZE 10
+static LRUEntry NULL_LRU_ENTRY; // Used to create an "empty" new LRUEntry
+
+LRUPolicy::~LRUPolicy()
+{
+  TSDebug(PLUGIN_NAME, "LRUPolicy DTOR");
+  TSMutexLock(_lock);
```
Comment (SolidWallOfCode) [inline plugins/cache_promote/lru_policy.cc:28]: Why are you locking in the destructor?  Is the destructor modifying other objects? If some other thread is trying to touch this and just misses getting the lock, then instead it'll crash working with a dead object.

File: plugins/cache_promote/chance_policy.h (hunk 1 above)
Comment (zwoop) [reply plugins/cache_promote/chance_policy.h:64]: Gah, this code is not part of the change :) Just moving code around.

File: plugins/cache_promote/lru_policy.h [hunk 3]
```diff
@@ -104,9 +104,16 @@ class LRUPolicy : public PromotionPolicy
     return "LRU";
   }
 
+  const std::string
+  id() const override
+  {
+    return _label + ";LRU=b:" + std::to_string(_buckets) + ",h:" + std::to_string(_hits);
```
Comment (SolidWallOfCode) [inline plugins/cache_promote/lru_policy.h:111]: Do these values change? Might set this up in the constructor and then return the member.

Comment (zwoop) [reply plugins/cache_promote/lru_policy.h:111]: No, those do not change, they are part of the LRU objects, the value are set and defined via the @pparam configurations. The label is used to make sure that we can modify the settings  in remap.config, such that the --label is the same, but e.g. the bucket size is changed.

I can save it in the constructor if you feel strongly, there's no performance impact here, but I see your point.

**Comment by zwoop** [comment]
Thanks Alan. Yeh, I should have pointed out, most of this change is refactoring the code, moving it out of a single monolithic C++ file, and into a modular design. I intentionally did not (and do not) want to modify that code, it's been like that for years (predating most of our C++11/17 migration).

**Comment by zwoop** [comment]
Argh, sorry, forgot to squash ...

**Comment by zwoop** [comment]
Cherry-picked to 8.1.x
Cherry-picked to v9.0.x branch.

**Pull request merged**